### PR TITLE
WIP: app refactoring

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

App refactoring PR based on #1459 with master merged in and post-merge fixes.
Pre-refactoring profiling showed there were lots of calls to memory allocations and GC possibly because of exceed copying.

### Refactoring goals:

* Raise application limits (program length, cost, number or keys) by reducing amount of memory operations
* Simplify the code

### Overview of changes:

1. `accountbase` table with serialized `AccountData` struct  does not have app state anymore. All the key-value pairs are in new `storage` table.
2. To reduce amount of data copied `ledger` has a new `FullLookup` function for returning full `AccountData` with app state and old `Lookup`, `LookupWithoutRewards` return `AccountData` without state.
3. To support 1 and 3 a new `miniAccountDelta` struct is widely used in ledger instead of `accountDelta`. This aims to speedup evaluation of regular transaction for accounts with lots of app state data.
4. App key-value logic moved from TEAL evaluator to ledger. This forced change of `LedgerForLogic` interface and rewriting balance adapter in dryrun REST handler and in TEAL debugger, and making `TxLease` struct exported.
5. Ledger's `eval` and `apply` relationship looks better after removal `StatefulEvaluator` interface used to break circular dependency.

## Test Plan

Recovered lots of tests and added new ones.

No noticeable performance gain with changes as is, need to profile and discover bottlenecks.

### CODE REVIEW

**Review by algonautshant — COMMENTED** [review]
I have some comments/questions.
Will continue with the review.

File: ledger/appcow.go [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+		calf.commitToParent()
+	}
+
+	return pass, evalDelta, nil
+}
+
+func (cb *roundCowState) BuildDelta(aidx basics.AppIndex, txn *transactions.Transaction) (evalDelta basics.EvalDelta, err error) {
+	foundGlobal := false
+	for addr, smod := range cb.mods.sdeltas {
+		for aapp, sdelta := range smod {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/appcow.go:464]: yes, one global and one local deltas per application. We need loop through apps anyway,

File: ledger/cow.go [hunk 2]
```diff
@@ -68,19 +91,27 @@ type StateDelta struct {
 	// last round for which we have seen a compact cert.
 	// zero if no compact cert seen.
 	compactCertSeen basics.Round
+
+	// previous block timestamp
+	prevTimestamp int64
+
+	// storage deltas
+	sdeltas map[basics.Address]map[storagePtr]*storageDelta
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/cow.go:99]: I agree

File: ledger/acctupdates.go [hunk 3]
```diff
@@ -1328,6 +1479,104 @@ func (au *accountUpdates) upgradeDatabaseSchema3(ctx context.Context, tx *sql.Tx
 	return 4, nil
 }
 
+func (au *accountUpdates) upgradeDatabaseSchema4(ctx context.Context, tx *sql.Tx) (updatedDBVersion int32, err error) {
```
Comment (algonautshant) [inline ledger/acctupdates.go:1482]: Please add a description of what is happening here.

File: ledger/accountdb.go [hunk 4]
```diff
@@ -119,6 +150,146 @@ type accountDelta struct {
...
+			}
+			owned[aapp.aidx] = states
+		}
+
+		data.AppLocalStates = owned
+	}
+	return data, nil
+}
+
+func withoutAppKV(u basics.AccountData) (res basics.AccountData) {
```
Comment (algonautshant) [inline ledger/accountdb.go:262]: What is this function returning?

File: ledger/acctupdates.go [hunk 5]
```diff
@@ -105,6 +105,42 @@ type modifiedCreatable struct {
 	ndeltas int
 }
 
+type modifiedStorage struct {
```
Comment (algonautshant) [inline ledger/acctupdates.go:108]: This file has grown too long. 
Maybe consider splitting it?

File: ledger/acctupdates.go [hunk 6]
```diff
@@ -1455,6 +1705,62 @@ func (au *accountUpdates) newBlockImpl(blk bookkeeping.Block, delta StateDelta)
 		au.creatables[cidx] = mcreat
 	}
 
+	for addr, smap := range delta.sdeltas {
+		for storagePtr, sdelta := range smap {
+			/*au.log.Warnf("MAXJ: %s", storagePtr.addr)
```
Comment (algonautshant) [inline ledger/acctupdates.go:1710]: Delete the comment here?

File: ledger/acctupdates.go [hunk 7]
```diff
@@ -1804,15 +2202,76 @@ func (au *accountUpdates) commitRound(offset uint64, dbRound basics.Round, lookb
...
+	storageKeyFlushcount := make(map[appKeyPtr]int)
 
 	var committedRoundDigest crypto.Digest
 
 	if isCatchpointRound {
 		committedRoundDigest = au.roundDigest[offset+uint64(lookback)-1]
 	}
 
+	// accountUpdates stores account and app deltas separately, but
+	// catchpoints expect unified deltas, keyed off of address.
```
Comment (algonautshant) [inline ledger/acctupdates.go:2215]: Can you comment on whether the catchpoints should also change to reflect the changes here?

File: ledger/ledger.go [hunk 8]
```diff
@@ -435,31 +451,43 @@ func (l *Ledger) ListApplications(maxAppIdx basics.AppIndex, maxResults uint64)
...
 	if err != nil {
 		return basics.AccountData{}, basics.Round(0), err
 	}
 
-	return data, validThrough, nil
+	return mini, validThrough, nil
+}
+
+// FullLookup is like Lookup but it includes application key-value pairs.
+func (l *Ledger) FullLookup(rnd basics.Round, addr basics.Address) (basics.AccountData, error) {
```
Comment (algonautshant) [inline ledger/ledger.go:487]: Rename this to LookupWithAppKV to be more informative?

**Review by algonautshant — COMMENTED** [review]
Some more comments...

**Comment by algorandskiy** [comment]
Replaced with #1765

**Pull request closed without merging**